OpenSim/Simulation/Model/ModelComponent.cpp

```diff
@@ -84,19 +96,38 @@ void ModelComponent::connectToModel(Model& model)
     extendConnectToModel(model);
 }
 
-// Base class implementation of virtual method.
-void ModelComponent::generateDecorations
-    (bool                                        fixed, 
-    const ModelDisplayHints&                    hints,
-    const SimTK::State&                         state,
-    SimTK::Array_<SimTK::DecorativeGeometry>&   appendToThis) const 
-{
-    for(unsigned int i=0; i < _components.size(); i++){
-        ModelComponent *mc = dynamic_cast<ModelComponent*>(_components[i]);
-        mc->generateDecorations(fixed,hints,state,appendToThis);
+void ModelComponent::adoptGeometry(OpenSim::Geometry* geom) {
+    // Check that name exists and is unique as it's used to form PathName
+    if (geom->getName().empty()){
+        bool nameFound = false;
+        int index = 1;
+        while (!nameFound){
+            std::stringstream ss;
+            // generate candiate name
+            ss << getName() << "_geom_" << index;
+            std::string candidate = ss.str();
+            bool exists = false;
+            for (int idx = 0; idx < getProperty_GeometrySet().size() && !exists; idx++){
+                if (get_GeometrySet(idx).getName() == candidate){
+                    exists = true;
+                    break;
+                }
+            }
+            if (!exists){
+                nameFound = true;
+                geom->setName(candidate);
+            }
+            else
+                index++;
+        }
+        
     }
+    append_GeometrySet(*geom);
+    addComponent(geom); 
```

Comment:
Ayman, I don't think this can be doing what you meant it to do. `append_GeometrySet()` takes a `Geometry` object (not a pointer) and pushes a copy of that object onto the property list. Then `addComponent()` takes a pointer and stores it, with no ownership transfer. So (1) the component is not the same object as is in the `GeometrySet` property, and (2) the heap-allocated object `geom` which is supposedly getting adopted ends up getting leaked instead since no one owns it. Also, any operations that are subsequently performed on the subcomponent don't affect the object stored in the GeometrySet property.